File: hdf5/fortran/hdf5_general.f90
 ! * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * !
 ! Copyright by The HDF Group.                                                   !
 ! All rights reserved.                                                          !
 !                                                                               !
 ! The full HDF5 copyright notice, including terms governing use, modification,  !
 ! and redistribution, is contained in the file COPYING.  COPYING can be found   !
 ! at the root of the source code distribution tree.                             !
 ! For questions contact the help desk at help@hdfgroup.org                      !
 ! * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * !
 !
 ! created by:  Isayah Reed
 !              May  4, 2011
 !
 ! Purpose:
 !  This is 1 of 3 simple HDF5 programs that demonstrates how to add CF attributes
 !  to an h5 file. Creates a file with 3 datasets: lat, lon, temp. Lat contains
 !  the CF attributes: units, long_name, and standard_name. Lon has the same CF
 !  attributes as the latitude dataset. Temp contains the CF attributes: units,
 !  long_name, _FillValue, coordinates, valid_min, valid_max, valid_range,
 !  scale_factor, add_offset. Outputs data to general.h5
 !  
 ! Compile:
 !  $(HDF5_DIR)/bin/h5fc hdf5_general.f90 -I$(HDF5_DIR)/include -L$(HDF5_DIR)/lib -lhdf5 -lhdf5_hl  -lhdf5hl_fortran -lhdf5_fortran -lz -lm
 !

PROGRAM main

  USE HDF5

  IMPLICIT NONE

  CHARACTER(LEN=10), PARAMETER :: filename = "general.h5"
  CHARACTER(LEN=4) , PARAMETER :: TEMP= "temp"
  CHARACTER(LEN=3) , PARAMETER :: LAT= "lat"
  CHARACTER(LEN=3) , PARAMETER :: LON= "lon"
  CHARACTER(LEN=9) , PARAMETER :: LONGNAME = "long_name"
  CHARACTER(LEN=5) , PARAMETER :: UNITS = "units"
  CHARACTER(LEN=10) , PARAMETER :: FILLVALUE = "_FillValue"
  CHARACTER(LEN=11) , PARAMETER :: COORDINATES = "coordinates"
  CHARACTER(LEN=11) , PARAMETER :: VALIDRANGE= "valid_range"
  CHARACTER(LEN=9) , PARAMETER :: VALIDMAX = "valid_max"
  CHARACTER(LEN=9) , PARAMETER :: VALIDMIN = "valid_min"
  CHARACTER(LEN=12) , PARAMETER :: SCALEFACTOR = "scale_factor"
  CHARACTER(LEN=10) , PARAMETER :: ADDOFFSET = "add_offset"
  CHARACTER(LEN=13) , PARAMETER :: STDNAME = "standard_name"

  INTEGER          , PARAMETER :: nx1     = 180
  INTEGER          , PARAMETER :: ny1     = 360

  INTEGER :: i, j, strlen, status
  INTEGER(HID_T) :: file, space, dset, attr_id, attr_space ! handles
  INTEGER(HID_T) :: atype_id, vtype_id
  CHARACTER(LEN=3), DIMENSION(1:2) ::  coorlist  ! coordinates {'lat', 'lon'}
  INTEGER(HSIZE_T), DIMENSION(1:2) :: temp_dims = (/ny1, nx1/) ! size read/write buffer
  INTEGER(HSIZE_T), DIMENSION(1:1) :: lat_dims = (/nx1/) 
  INTEGER(HSIZE_T), DIMENSION(1:1) :: lon_dims = (/ny1/) 
  INTEGER(HSIZE_T), DIMENSION(1)  :: dimsf ! dimension for scalars
  INTEGER(HSIZE_T), DIMENSION(2) :: dimsa   ! dimensions for arrays
  DOUBLE PRECISION, DIMENSION(1:nx1,1:ny1) :: temp_data
  DOUBLE PRECISION, DIMENSION(nx1) :: lat_data
  DOUBLE PRECISION, DIMENSION(ny1) :: lon_data
  DOUBLE PRECISION :: value
  DOUBLE PRECISION, DIMENSION(1:2) :: arrdata 
  CHARACTER(LEN=20) :: strdata
  CHARACTER, DIMENSION(2) ::  hvl_data


  ! Initialize FORTRAN interface.
  CALL h5open_f(status)
  
  ! Initialize temperature data.
  DO i = 1, 60
     DO j = 1, ny1 
        temp_data(i,j) = 280.0
     ENDDO
  ENDDO
  DO i = 61, 120
     DO j = 1, ny1 
        temp_data(i,j) = 300.0
     ENDDO
  ENDDO
  DO i = 121, nx1
     DO j = 1, ny1 
        temp_data(i,j) = 280.0
     ENDDO
  ENDDO
  
  ! Create a new file using the default properties.
  CALL h5fcreate_f(filename, H5F_ACC_TRUNC_F, file, status)


  ! PART 1: temperature

  ! Create dataspace.  Setting size to be the current size of the 
  ! temperature array
  CALL h5screate_simple_f(2, temp_dims, space, status)
  
  ! Create the dataset.  We will use all default properties for this
  ! example.
  CALL h5dcreate_f(file, TEMP, h5t_ieee_f32le, space, dset, status)
  
  ! Write the data to the dataset.
  CALL h5dwrite_f(dset, H5T_NATIVE_DOUBLE, temp_data, temp_dims, status)
  CALL h5sclose_f(space, status)
  
  ! add 1st attribute: units
  strdata= "kelvin"
  strlen= 6
  ! Create scalar data space for the attribute.
  CALL h5screate_f(H5S_SCALAR_F, space, status)
  ! Create datatype for the attribute.
  CALL h5tcopy_f(H5T_NATIVE_CHARACTER, atype_id, status)
  ! set the size of the datatype to be the size of the string "kelvin"
  CALL h5tset_size_f(atype_id, 6, status)
  ! set string as NULLTERM to imitate C-style strings
  CALL h5tset_strpad_f(atype_id, H5T_STR_NULLTERM_F, status)
  !Create dataset attribute.
  CALL h5acreate_f(dset, UNITS, atype_id, space, &
                      attr_id, status)
  ! Write the attribute data.
  dimsf = 1
  CALL h5awrite_f(attr_id, atype_id, "kelvin", dimsf, status)
  ! Close the dataspace
  CALL h5sclose_f(space, status)
  ! Close the attribute.
  CALL h5aclose_f(attr_id, status)

  ! add 2nd attribute: fillvalue
  dimsf = 1
  value= -999.0
  CALL h5screate_f(H5S_SCALAR_F, space, status)
  CALL h5tcopy_f(h5t_ieee_f32le, atype_id, status)
  CALL h5acreate_f(dset, FILLVALUE, atype_id, space, &
                      attr_id, status)
  CALL h5awrite_f(attr_id, H5T_NATIVE_DOUBLE, value, dimsf, status)
  CALL h5sclose_f(space, status)
  CALL h5aclose_f(attr_id, status)

  ! add 3rd attribute: long name
  dimsf = 1
  strdata= "temperature"
  strlen= 11
  CALL h5screate_f(H5S_SCALAR_F, space, status)
  CALL h5tcopy_f(H5T_NATIVE_CHARACTER, atype_id, status)
  CALL h5tset_size_f(atype_id, strlen, status)
  CALL h5tset_strpad_f(atype_id, H5T_STR_NULLTERM_F, status)
  CALL h5acreate_f(dset, LONGNAME, atype_id, space, &
                      attr_id, status)
  CALL h5awrite_f(attr_id, atype_id, strdata, dimsf, status)
  CALL h5sclose_f(space, status)
  CALL h5aclose_f(attr_id, status)

  ! add 4th attribute: coordinates
  dimsf = 2
  coorlist(1)= "lat"
  coorlist(2)= "lon"
  strlen= 3
  CALL h5screate_simple_f(1, dimsf, space, status)
  CALL h5tcopy_f(H5T_NATIVE_CHARACTER, atype_id, status)
  CALL h5tset_size_f(atype_id, strlen, status)
  CALL h5tset_strpad_f(atype_id, H5T_STR_NULLTERM_F, status)
  CALL h5acreate_f(dset, COORDINATES, atype_id, space, &
                      attr_id, status)
  CALL h5awrite_f(attr_id, atype_id, coorlist, dimsf, status)
  CALL h5sclose_f(space, status)
  CALL h5aclose_f(attr_id, status)

  ! add attribute: add_offset
  dimsf = 1
  value= 5
  CALL h5screate_f(H5S_SCALAR_F, space, status)
  CALL h5tcopy_f(h5t_ieee_f32le, atype_id, status)
  CALL h5acreate_f(dset, ADDOFFSET, atype_id, space, &
                      attr_id, status)
  CALL h5awrite_f(attr_id, H5T_NATIVE_DOUBLE, value, dimsf, status)
  CALL h5sclose_f(space, status)
  CALL h5aclose_f(attr_id, status)

  ! add attribute: SCALEFACTOR
  dimsf = 1
  value= 1.01
  CALL h5screate_f(H5S_SCALAR_F, space, status)
  CALL h5tcopy_f(h5t_ieee_f32le, atype_id, status)
  CALL h5acreate_f(dset, SCALEFACTOR, atype_id, space, &
                      attr_id, status)
  CALL h5awrite_f(attr_id, H5T_NATIVE_DOUBLE, value, dimsf, status)
  CALL h5sclose_f(space, status)
  CALL h5aclose_f(attr_id, status)


  ! add attribute: valid_min
  dimsf = 1
  value= 0
  CALL h5screate_f(H5S_SCALAR_F, space, status)
  CALL h5tcopy_f(h5t_ieee_f32le, atype_id, status)
  CALL h5acreate_f(dset, VALIDMIN, atype_id, space, &
                      attr_id, status)
  CALL h5awrite_f(attr_id, H5T_NATIVE_DOUBLE, value, dimsf, status)
  CALL h5sclose_f(space, status)
  CALL h5aclose_f(attr_id, status)

  ! add attribute: valid_max
  dimsf = 1
  value= 400
  CALL h5screate_f(H5S_SCALAR_F, space, status)
  CALL h5tcopy_f(h5t_ieee_f32le, atype_id, status)
  CALL h5acreate_f(dset, VALIDMAX, atype_id, space, &
                      attr_id, status)
  CALL h5awrite_f(attr_id, H5T_NATIVE_DOUBLE, value, dimsf, status)
  CALL h5sclose_f(space, status)
  CALL h5aclose_f(attr_id, status)

  ! add attribute: valid_range
  dimsf = 2
  arrdata(1)= 9
  arrdata(2)= 400
  CALL h5screate_simple_f(1, dimsf, space, status)
  CALL h5tcopy_f(h5t_ieee_f32le, atype_id, status)
  !CALL h5tset_size_f(atype_id, H5T_VLEN_F, status)
  CALL h5acreate_f(dset, VALIDRANGE, atype_id, space, &
                      attr_id, status)
  CALL h5awrite_f(attr_id, atype_id, arrdata, dimsf, status)
  CALL h5sclose_f(space, status)
  CALL h5aclose_f(attr_id, status)

  CALL h5dclose_f(dset, status)
  

  ! PART 2: latitude

  ! initialize data
  lat_data(1)= -90.0
  DO i = 2, nx1
    lat_data(i) = lat_data(i-1)+1.0
  ENDDO

  CALL h5screate_simple_f(1, lat_dims, space, status)
  CALL h5dcreate_f(file, LAT, h5t_ieee_f32le, space, dset, status)
  CALL h5dwrite_f(dset, H5T_NATIVE_DOUBLE, lat_data, lat_dims, status)
  CALL h5sclose_f(space, status)

  ! add 1st attribute: long_name
  dimsf = 1
  strdata= "latitude"
  strlen= 8
  CALL h5screate_f(H5S_SCALAR_F, space, status)
  CALL h5tcopy_f(H5T_NATIVE_CHARACTER, atype_id, status)
  CALL h5tset_size_f(atype_id, strlen, status)
  CALL h5tset_strpad_f(atype_id, H5T_STR_NULLTERM_F, status)
  CALL h5acreate_f(dset, LONGNAME, atype_id, space, &
                      attr_id, status)
  dimsf = 2
  CALL h5awrite_f(attr_id, atype_id, strdata, dimsf, status)
  CALL h5sclose_f(space, status)
  CALL h5aclose_f(attr_id, status)

  ! add 2nd attribute: units
  dimsf = 1
  strdata= "degrees_north"
  strlen= 13
  CALL h5screate_f(H5S_SCALAR_F, space, status)
  CALL h5tcopy_f(H5T_NATIVE_CHARACTER, atype_id, status)
  CALL h5tset_size_f(atype_id, strlen, status)
  CALL h5tset_strpad_f(atype_id, H5T_STR_NULLTERM_F, status)
  CALL h5acreate_f(dset, UNITS, atype_id, space, &
                      attr_id, status)
  dimsf = 2
  CALL h5awrite_f(attr_id, atype_id, strdata, dimsf, status)
  CALL h5sclose_f(space, status)
  CALL h5aclose_f(attr_id, status)

  ! add standard name
  dimsf = 1
  strdata= "latitude"
  strlen= 8
  CALL h5screate_f(H5S_SCALAR_F, space, status)
  CALL h5tcopy_f(H5T_NATIVE_CHARACTER, atype_id, status)
  CALL h5tset_size_f(atype_id, strlen, status)
  CALL h5tset_strpad_f(atype_id, H5T_STR_NULLTERM_F, status)
  CALL h5acreate_f(dset, STDNAME, atype_id, space, &
                      attr_id, status)
  CALL h5awrite_f(attr_id, atype_id, strdata, dimsf, status)
  CALL h5sclose_f(space, status)
  CALL h5aclose_f(attr_id, status)

  CALL h5dclose_f(dset, status)

  ! PART 3: longitude

  ! initialize data
  lon_data(1)= -180.0
  DO i = 2, ny1
    lon_data(i) = lon_data(i-1)+1.0
  ENDDO

  CALL h5screate_simple_f(1, lon_dims, space, status)
  CALL h5dcreate_f(file, LON, h5t_ieee_f32le, space, dset, status)
  CALL h5dwrite_f(dset, H5T_NATIVE_DOUBLE, lon_data, lon_dims, status)
  CALL h5sclose_f(space, status)

  ! add 1st attribute: long_name
  dimsf = 1
  strdata= "longitude"
  strlen= 9
  CALL h5screate_f(H5S_SCALAR_F, space, status)
  CALL h5tcopy_f(H5T_NATIVE_CHARACTER, atype_id, status)
  CALL h5tset_size_f(atype_id, strlen, status)
  CALL h5tset_strpad_f(atype_id, H5T_STR_NULLTERM_F, status)
  CALL h5acreate_f(dset, LONGNAME, atype_id, space, &
                      attr_id, status)
  dimsf = 2
  CALL h5awrite_f(attr_id, atype_id, strdata, dimsf, status)
  CALL h5sclose_f(space, status)
  CALL h5aclose_f(attr_id, status)

  ! add 2nd attribute: units
  dimsf = 1
  strdata= "degrees_east"
  strlen= 12
  CALL h5screate_f(H5S_SCALAR_F, space, status)
  CALL h5tcopy_f(H5T_NATIVE_CHARACTER, atype_id, status)
  CALL h5tset_size_f(atype_id, strlen, status)
  CALL h5tset_strpad_f(atype_id, H5T_STR_NULLTERM_F, status)
  CALL h5acreate_f(dset, UNITS, atype_id, space, &
                      attr_id, status)
  dimsf = 2
  CALL h5awrite_f(attr_id, atype_id, strdata, dimsf, status)
  CALL h5sclose_f(space, status)
  CALL h5aclose_f(attr_id, status)

  ! add standard name
  dimsf = 1
  strdata= "longitude"
  strlen= 9
  CALL h5screate_f(H5S_SCALAR_F, space, status)
  CALL h5tcopy_f(H5T_NATIVE_CHARACTER, atype_id, status)
  CALL h5tset_size_f(atype_id, strlen, status)
  CALL h5tset_strpad_f(atype_id, H5T_STR_NULLTERM_F, status)
  CALL h5acreate_f(dset, STDNAME, atype_id, space, &
                      attr_id, status)
  CALL h5awrite_f(attr_id, atype_id, strdata, dimsf, status)
  CALL h5sclose_f(space, status)
  CALL h5aclose_f(attr_id, status)

  CALL h5dclose_f(dset , status)

  ! Close and release resources.
  CALL h5fclose_f(file , status)
  !
END PROGRAM main

File: hdf5/c/hdf5_general.c
/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * !
 * Copyright by The HDF Group.                                                   !
 * All rights reserved.                                                          !
 *                                                                               !
 * This code is provided as open source but it NOT licensed, has no limitations, !
 * and has no expressed or implied warranties.
 * For questions contact the help desk at help@hdfgroup.org                      !
 * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */

/*
 * created by:  Isayah Reed
 *              May  4, 2011
 *
 * Purpose: 
 *  This is 1 of 4 simple HDF5 programs that demonstrates how to add CF attributes 
 *  to an h5 file. Creates a file with 3 datasets: lat, lon, temp. Lat contains 
 *  the CF attributes: units, long_name, and standard_name. Lon has the same CF 
 *  attributes as the latitude dataset. Temp contains the CF attributes: units, 
 *  long_name, _FillValue, coordinates, valid_min, valid_max, valid_range, 
 *  scale_factor, add_offset. Outputs data to general.h5 
 *
 * Compile:
 *  gcc hdf5_general.c -I$(HDF5_DIR)/include -L$(HDF5_DIR)/lib -lhdf5_hl -lhdf5 -lz -lm
 *
*/


#include "hdf5.h"
#include "string.h"
#include <time.h>

#define H5FILE_NAME "general.h5"
/* dataset names */
#define TEMP "temp"
#define LAT "lat"
#define LON "lon"
/* CF attributes */
#define UNITS "units"
#define FILLVALUE "_FillValue"
#define LONGNAME "long_name"
#define COORDINATES "coordinates"
#define VALIDMIN "valid_min"
#define VALIDMAX "valid_max"
#define VALIDRANGE "valid_range"
#define SCALEFACTOR "scale_factor"
#define ADDOFFSET "add_offset"
#define STDNAME "standard_name"



int main(void)
{
    hid_t file, dataset, attr;	/* file, dataset, attribute handles */
    hid_t floatType, stringType, arrayType;	/* datatypes */
    hid_t floatSpace, stringSpace, arraySpace;	/* dataspaces */
    hsize_t dimsa[2], dimsf;	/* dataset dimensions */
    herr_t status;
    int i, j;
    float temp_array[180][360], lat_array[180], lon_array[360], value[2];
    char *coorlist[2] = { "lat", "lon" };

    int xrand[30], yrand[30]; /* array to store the random fillvalues */
    srand(time(NULL));


    file = H5Fcreate(H5FILE_NAME, H5F_ACC_TRUNC, H5P_DEFAULT, H5P_DEFAULT);

/* PART 1: TEMPERATURE */

    /* values between a[60][*] and a[120][*] is around 300.0 */
    for(i = 60; i <= 120; i++)
	    for(j = 0; j < 360; j++)
	        temp_array[i][j] = 300.0;

    /* values between a[0][*] and a[59][*], a[121][*] and a[179][*] 
       is around 280.0 */
    for(i = 0; i < 60; i++)
	    for(j = 0; j < 360; j++)
	        temp_array[i][j] = 280.0;

    for(i = 121; i < 180; i++)
	    for(j = 0; j < 360; j++)
	        temp_array[i][j] = 280.0;

/* UNCOMMENT TO ADD RANDOM FILLVALUES TO THE TEMPERATURE DATASET */
    /* a few random values should be -999.0  */
/*
    for(i=0; i<30; i++)
        xrand[i]= rand()%180;
    for(i=0; i<30; i++)
        yrand[i]= rand()%360;
*/
    /* insert the fillvalue (-999) in random parts of the temp array */
/*
    for(i=0; i<30; i++)
        temp_array[xrand[i]][yrand[i]]= -999.0;
*/


    /* initialize handles */
    floatType = H5Tcopy(H5T_NATIVE_FLOAT);
    H5Tset_precision(floatType, 32); /* set as 32bit */
    dimsa[0] = 180;
    dimsa[1] = 360;
    floatSpace = H5Screate_simple(2, dimsa, NULL);
    stringType = H5Tcopy(H5T_C_S1);
    stringSpace = H5Screate(H5S_SCALAR);

    /* create the temperature dataset  */
    dataset = H5Dcreate(file, TEMP, floatType, floatSpace,
			H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);
    status = H5Dwrite(dataset, H5T_NATIVE_FLOAT, H5S_ALL, H5S_ALL,
		      H5P_DEFAULT, temp_array);


    /* add UNITS attribute */
    /* UNITS: a string that represents the quantity of measurement */
    status = H5Tset_size(stringType, (hsize_t) strlen("kelvin"));
    attr= H5Acreate(dataset, UNITS, stringType, stringSpace, 
            H5P_DEFAULT, H5P_DEFAULT);
    status = H5Awrite(attr, stringType, "kelvin");
    H5Aclose(attr);

    /* add FILLVALUE attribute */
    /* FILLVALUE: represents a missing or undefined value */
    dimsf = 1;
    value[0] = -999.0;
    floatSpace = H5Screate_simple(1, &dimsf, NULL);
    attr= H5Acreate(dataset, FILLVALUE, floatType, floatSpace, 
            H5P_DEFAULT, H5P_DEFAULT);
    status = H5Awrite(attr, floatType, &value[0]);
    H5Aclose(attr);

    /* add LONGNAME attribute */
    /* LONGNAME: a string representing a long descriptive name for the data */
    status = H5Tset_size(stringType, (hsize_t) strlen("temperature"));
    attr= H5Acreate(dataset, LONGNAME, stringType, stringSpace, 
            H5P_DEFAULT, H5P_DEFAULT);
    status = H5Awrite(attr, stringType, "temperature");
    H5Aclose(attr);

    /* add COORDINATES attribute */
    /* COORDINATES: a string list of the associated coordinate variable 
          names of the variable */
    dimsa[0] = 2;
    arraySpace = H5Screate_simple(1, &dimsa[0], NULL);
    arrayType = H5Tcopy(H5T_C_S1);
    /* set the size of each string entry to be unlimited length */
    status = H5Tset_size(arrayType, H5T_VARIABLE);
    attr= H5Acreate(dataset, COORDINATES, arrayType, arraySpace,
		    H5P_DEFAULT, H5P_DEFAULT);
    status = H5Awrite(attr, arrayType, coorlist);
    H5Aclose(attr);

    /* add VALID_MIN attribute */
    /* VALID_MIN: smallest valid value of a variable */
    value[0]= 0.0;
    floatSpace= H5Screate(H5S_SCALAR);
    attr= H5Acreate(dataset, VALIDMIN, floatType, floatSpace, 
            H5P_DEFAULT, H5P_DEFAULT);
    status= H5Awrite(attr,floatType,&value[0]);
    H5Sclose(floatSpace);
    H5Aclose(attr);

    /* add VALID_MAX attribute */
    /* VALID_MAX: Smallest valid value of a variable */
    value[0]= 400.0;
    floatSpace= H5Screate(H5S_SCALAR);
    attr= H5Acreate(dataset, VALIDMAX, floatType, floatSpace, 
            H5P_DEFAULT, H5P_DEFAULT);
    status= H5Awrite(attr,floatType,&value[0]);
    H5Sclose(floatSpace);
    H5Aclose(attr);


    /*NOTE: a dataset should not contain both VALID_RANGE and VALID_MIN/MAX */


    /* add VALID_RANGE attribute */
    /* VALID_RANGE: smallest and largest valid values of a variable */
    dimsf= 2;
    arraySpace= H5Screate_simple(1, &dimsf, NULL);
    value[0]= 275;
    value[1]= 305;
    attr= H5Acreate(dataset, VALIDRANGE, floatType, arraySpace,
        H5P_DEFAULT, H5P_DEFAULT);
    status= H5Awrite(attr,floatType,value);
    H5Sclose(arraySpace);
    H5Aclose(attr);

    /* add ADD_OFFSET attribute */
    /* ADD_OFFSET: number added to the data after it is read by an application */
    value[0]= 1000.0;
    floatSpace= H5Screate(H5S_SCALAR);
    attr= H5Acreate(dataset, ADDOFFSET, floatType, floatSpace, 
            H5P_DEFAULT, H5P_DEFAULT);
    status= H5Awrite(attr,floatType,&value[0]);
    H5Sclose(floatSpace);
    H5Aclose(attr);

    /* add SCALE_FACTOR attribute */
    /* SCALE_FACTOR: factor to multiply after data is read */
    value[0]= 5.0;
    floatSpace= H5Screate(H5S_SCALAR);
    attr= H5Acreate(dataset, SCALEFACTOR, floatType, floatSpace, 
            H5P_DEFAULT, H5P_DEFAULT);
    status= H5Awrite(attr,floatType,&value[0]);
    H5Sclose(floatSpace);
    H5Aclose(attr);


    H5Dclose(dataset);



/**** PART 2: LATITUDE ****/

    /*initialize data array: arithmetic sequence from -90.0 - 89.0 */
    lat_array[0] = -90.0;
    for (i = 1; i < 180; i++)
	lat_array[i] = lat_array[i - 1] + 1.0;

    dimsf = 180;
    floatSpace = H5Screate_simple(1, &dimsf, NULL);


    dataset = H5Dcreate(file, LAT, floatType, floatSpace,
			H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);
    status = H5Dwrite(dataset, H5T_NATIVE_FLOAT, H5S_ALL, H5S_ALL,
		      H5P_DEFAULT, lat_array);

    /* add LONG_NAME attribute */
    /*stringSpace= H5Screate(H5S_SCALAR); */
    status = H5Tset_size(stringType, (hsize_t) strlen("latitude"));
    attr= H5Acreate(dataset, LONGNAME, stringType, stringSpace, 
            H5P_DEFAULT, H5P_DEFAULT);
    status = H5Awrite(attr, stringType, "latitude");
    H5Aclose(attr);

    /* add UNITS attribute */
    status = H5Tset_size(stringType, (hsize_t) strlen("degrees_north"));
    attr= H5Acreate(dataset, UNITS, stringType, stringSpace, 
            H5P_DEFAULT, H5P_DEFAULT);
    status = H5Awrite(attr, stringType, "degrees_north");
    H5Aclose(attr);

    /* add STANDARD_NAME attribute */
    /* STANDARD_NAME: identifies the variable is the coordinate variable */
    status= H5Tset_size(stringType, (hsize_t)strlen("latitude"));
    attr= H5Acreate(dataset, STDNAME, stringType, stringSpace,
            H5P_DEFAULT, H5P_DEFAULT);
    status= H5Awrite(attr,stringType,"latitude");
    H5Aclose(attr);

    H5Dclose(dataset);




/**** Part 3: LONGITUDE ****/

    /*initialize: arithmetic sequence from -180.0 - 179.0 */
    lon_array[0] = -180.0;
    for (i = 1; i < 360; i++)
	lon_array[i] = lon_array[i - 1] + 1.0;

    dimsf = 360;
    floatSpace = H5Screate_simple(1, &dimsf, NULL);

    dataset = H5Dcreate(file, LON, floatType, floatSpace,
			H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);
    status = H5Dwrite(dataset, H5T_NATIVE_FLOAT, H5S_ALL, H5S_ALL,
		      H5P_DEFAULT, lon_array);

    /* add LONGNAME attribute */
    status = H5Tset_size(stringType, (hsize_t) strlen("longitude"));
    attr= H5Acreate(dataset, LONGNAME, stringType, stringSpace, 
            H5P_DEFAULT, H5P_DEFAULT);
    status = H5Awrite(attr, stringType, "longitude");
    H5Aclose(attr);

    /* add UNITS attribute */
    status = H5Tset_size(stringType, (hsize_t) strlen("degrees_east"));
    attr= H5Acreate(dataset, UNITS, stringType, stringSpace, 
            H5P_DEFAULT, H5P_DEFAULT);
    status = H5Awrite(attr, stringType, "degrees_east");
    H5Aclose(attr);

    /* add STANDARD_NAME attribute */
    /* STANDARD_NAME: identifies the variable is the coordinate variable */
    status= H5Tset_size(stringType, (hsize_t)strlen("longitude"));
    attr= H5Acreate(dataset, STDNAME, stringType, stringSpace,
            H5P_DEFAULT, H5P_DEFAULT);
    status= H5Awrite(attr,stringType,"longitude");
    H5Aclose(attr);
    H5Dclose(dataset);

    /* Close/release resources. */
    H5Sclose(floatSpace);
    H5Sclose(stringSpace);
    H5Sclose(arraySpace);
    H5Tclose(floatType);
    H5Tclose(stringType);
    H5Tclose(arrayType);
    H5Fclose(file);

    return 0;
}
